src/clients/Client.cc

```diff
@@ -759,8 +757,14 @@ Client::resumeBodyStorage()
 
     handleMoreAdaptedBodyAvailable();
 
-    if (adaptedBodySource != nullptr && adaptedBodySource->exhausted())
-        endAdaptedBodyConsumption();
+    if (adaptedBodySource != nullptr && adaptedBodySource->exhausted()) {
+        if (receivedWholeAdaptedReply || adaptedReplyAborted) {
+            // noteBodyProductionEnded() or noteBodyProducerAborted() could not endAdaptedBodyConsumption()
+            // while waiting for resumeBodyStorage() to consume buffered adapted body data.
+            endAdaptedBodyConsumption();
+        }
+        // else wait for noteBodyProductionEnded() or noteBodyProducerAborted()
+    }
 }
```

Comment:
Currently, there are three methods where Client reaches handleAdaptationCompleted(): resumeBodyStorage(), handleAdaptedBodyProductionEnded(), and handleAdaptedBodyProducerAborted(). All three methods have slightly different handleAdaptationCompleted()-path preconditions. Only two of them call endAdaptedBodyConsumption(). I am worried that we will eventually forget to update one of the three places when refactoring the other two.

Please try to reduce those differences by replacing endAdaptedBodyConsumption() with checkAdaptationCompletion() method that is called unconditionally from all three methods. The new method may start with an `if` statement checking for source exhaustion (like the one in handleAdaptedBodyProducerAborted()), proceed with (receivedWholeAdaptedReply or adaptedReplyAborted) check (like the one above), and, if none of those `if` statements called `return`, end with the current endAdaptedBodyConsumption() code.

Implementing this requires moving handledEarlyAdaptationAbort() call in handleAdaptedBodyProducerAborted() up, before the checkAdaptationCompletion() call. I hope that move is safe, but please double check. Besides code analysis, there might be more info about it in the commit(s) adding or moving that handledEarlyAdaptationAbort() call.

Reply:
Done. 

> moving handledEarlyAdaptationAbort() call in handleAdaptedBodyProducerAborted()

After this move, handledEarlyAdaptationAbort() is called before stopConsumingFrom(adaptedBodySource). I could not find in 7224ca5 any explanation about the old methods call order. handledEarlyAdaptationAbort() returning true means that it "fully handled adaptation abort", including the final abortAll(). All required cleanup will be performed in the follow-up calls, Client::swanSong() and cleanAdaptation(). So it looks like that handledEarlyAdaptationAbort() does not need the precondition of stopConsumingFrom() call. Also Client::handleAdaptationBlocked() does almost the same as handledEarlyAdaptationAbort() at its end and there is no stopConsumingFrom() there. So I agree that this move should be safe.